Fix for #1553

loganmc10 opened the pull request:
The problem currently is that these shaders are created when CombinerProgramBuilder is created, but g_cycleType isn't updated until later, however the shader text has an if statement that depends on the value of g_cycleType.

I moved the creation of the shader text to an "update()" function that is called right before the shader is written. This fixed the issue for me.

gonetz commented:
It was the first task in my todo list. Great that you found the problem, thanks! I have few notes about implementation though.

gonetz commented on src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilder.cpp:
```diff
@@ -539,14 +544,21 @@ class ShaderBlender1 : public ShaderPart
 	ShaderBlender2(const opengl::GLInfo & _glinfo)
 	{
-		if (_glinfo.isGLES2) {
+		m_glinfo = _glinfo;
```
ShaderBlender2 does not use g_cycleType. These changes are redundant.

gonetz commented on src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilder.cpp:
```diff
@@ -539,14 +544,21 @@ class ShaderBlender1 : public ShaderPart
 		}
 	}
+private:
+	opengl::GLInfo m_glinfo;
```
You need only _glinfo.isGLES2.
So,
`bool m_isGLES2;`

gonetz commented on src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilder.cpp:
```diff
@@ -466,7 +466,12 @@ class ShaderBlender1 : public ShaderPart
 	ShaderBlender1(const opengl::GLInfo & _glinfo)
 	{
-		if (_glinfo.isGLES2) {
+		m_glinfo = _glinfo;
```
`m_isGLES2 = _glinfo.isGLES2;`
see below for m_isGLES2

gonetz replied:
Also, I would keep current value of g_cycleType:
m_cycleType = g_cycleType;

gonetz commented on src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilder.cpp:
```diff
@@ -466,7 +466,12 @@ class ShaderBlender1 : public ShaderPart
+
+	void update()
+	{
+		if (m_glinfo.isGLES2) {
```
This will update m_part each time new shader is created.
I suggest to add this check:
```
if (m_cycleType == g_cycleType)
  return;
```

gonetz replied:
if (m_isGLES2)

gonetz commented:
Another idea: check both blenders for isGLESX, not for isGLES2. retrobenny said that Android devices have performance issues with new blending at high resolutions, so this optimization can be useful for all mobile platforms.

gonetz commented:
Another idea: add another class member
`std::string m_part2;`
and create two versions of shader: for 1cycle and 2cycle modes. Make ShaderPart::write virtual and override it for ShaderBlender1. It will be faster than repeatedly rewrite m_part. 
```
void ShaderBlender1::write(std::stringstream & shader) const override
{
  if (g_cycleType == G_CYC_1CYCLE)
	shader << m_part;
  else
	shader << m_part2;
}

```

gonetz commented:
Also, please make more detailed commit description.

loganmc10 commented:
You can commit a fix for this if you'd like and just close this PR, the problem is known now, I don't mind if you just fix it as you see best.

> Another idea: check both blenders for isGLESX, not for isGLES2. retrobenny said that Android devices have performance issues with new blending at high resolutions, so this optimization can be useful for all mobile platforms.

retrobenny normally uses a Shield TV, which doesn't use GLES anyway, but either way, I wouldn't really call it an "optimization". It was put in place to work around this issue: https://github.com/gonetz/GLideN64/issues/982 . It's just 2 ways of writing the same thing, but I would think it's best to prefer the shorter version if it works and let the GPU's GLSL compiler optimize it as best it can

gonetz commented:
Ok.

gonetz commented:
> I wouldn't really call it an "optimization".

There was a discussion about this matter. I don't remember details, but main thought is that GPUs do not like variable as index for vertex/matrix even if they support such thing. Some shader compilers produce low performance warnings. So, your code is kind of optimization.

loganmc10 commented:
> There was a discussion about this matter. I don't remember details, but main thought is that GPUs do not like variable as index for vertex/matrix even if they support such thing. Some shader compilers produce low performance warnings. So, your code is kind of optimization.

It's possible, I'm sure it wouldn't really hurt to enable it for all GLES devices

loganmc10 commented:
@gonetz I've updated this PR with the changes you suggested. I also removed the original blending implementation and made the GLES2/Raspberry Pi version the default. It produces the same result, and from retrobenny's testing it sounds like it does perform better.

gonetz commented:
@loganmc10 I already re-implemented your fix. It affects the same places of code, which I touched during my work on F-Zero, thus I put it on top to avoid conflicts. It will be released within days. Sorry, I did not informed you about it.

Regarding original blending implementation : yes, retrobenny says that your optimized version works almost as fast as legacy blending. However I will not completely remove it from code. I'll disable it by #ifdef and keep as reference, because it has much more simple and clear form.

loganmc10 commented:
Oh ok no problem, I'll close this then.
